File: js/src/tls/sig_algorithms.js
/**
 * PQC Signature Algorithms for TLS 1.3.
 *
 * Defines PQC and composite signature algorithms for the `signature_algorithms`
 * extension (CertificateVerify), along with sign/verify helpers.
 */

import { keyGen as mldsaKeyGen, sign as mldsaSign, verify as mldsaVerify } from '../mldsa/dsa.js';
import { ML_DSA_44, ML_DSA_65, ML_DSA_87 } from '../mldsa/params.js';
import {
  MLDSA65_ED25519, MLDSA87_ED25519,
  keyGen as compositeKeyGen, sign as compositeSign, verify as compositeVerify,
} from '../composite/composite_sig.js';

/** TLS 1.3 signature algorithm code points for PQC. */
export const SignatureAlgorithm = Object.freeze({
  MLDSA44:         0x0904,
  MLDSA65:         0x0905,
  MLDSA87:         0x0906,
  MLDSA65_ED25519: 0x0907,
  MLDSA87_ED25519: 0x0908,
});

/** All defined signature algorithms. */
export const ALL_SIGNATURE_ALGORITHMS = Object.freeze([
  SignatureAlgorithm.MLDSA44,
  SignatureAlgorithm.MLDSA65,
  SignatureAlgorithm.MLDSA87,
  SignatureAlgorithm.MLDSA65_ED25519,
  SignatureAlgorithm.MLDSA87_ED25519,
]);

/** Human-readable name for a signature algorithm. */
export function signatureAlgorithmName(algId) {
  switch (algId) {
    case SignatureAlgorithm.MLDSA44: return 'MLDSA44';
    case SignatureAlgorithm.MLDSA65: return 'MLDSA65';
    case SignatureAlgorithm.MLDSA87: return 'MLDSA87';
    case SignatureAlgorithm.MLDSA65_ED25519: return 'MLDSA65_ED25519';
    case SignatureAlgorithm.MLDSA87_ED25519: return 'MLDSA87_ED25519';
    default: return 'Unknown';
  }
}

/** Look up a signature algorithm by code point. */
export function signatureAlgorithmFromCodePoint(cp) {
  if (ALL_SIGNATURE_ALGORITHMS.includes(cp)) return cp;
  return undefined;
}

/** Whether the algorithm is a composite (hybrid) signature. */
export function isComposite(algId) {
  return algId === SignatureAlgorithm.MLDSA65_ED25519 ||
         algId === SignatureAlgorithm.MLDSA87_ED25519;
}

function mldsaParams(algId) {
  switch (algId) {
    case SignatureAlgorithm.MLDSA44: return ML_DSA_44;
    case SignatureAlgorithm.MLDSA65: return ML_DSA_65;
    case SignatureAlgorithm.MLDSA87: return ML_DSA_87;
    default: throw new Error('Not a pure ML-DSA algorithm');
  }
}

function compositeScheme(algId) {
  switch (algId) {
    case SignatureAlgorithm.MLDSA65_ED25519: return MLDSA65_ED25519;
    case SignatureAlgorithm.MLDSA87_ED25519: return MLDSA87_ED25519;
    default: throw new Error('Not a composite algorithm');
  }
}

/**
 * Generate a signing key pair.
 *
 * For composite algorithms, the returned object includes a `_compositeKP`
 * property containing the full composite key pair (including JCA key objects
 * needed for signing).
 *
 * @param {number} algId - Signature algorithm code point
 * @returns {{ pk: Uint8Array, sk: Uint8Array, algorithm: number, _compositeKP?: object }}
 */
export function generateSigningKey(algId) {
  if (isComposite(algId)) {
    const scheme = compositeScheme(algId);
    const kp = compositeKeyGen(scheme);
    return { pk: kp.pk, sk: kp.sk, algorithm: algId, _compositeKP: kp };
  }
  const params = mldsaParams(algId);
  const { pk, sk } = mldsaKeyGen(params);
  return { pk, sk, algorithm: algId };
}

/**
 * Sign a TLS 1.3 CertificateVerify handshake hash.
 *
 * For composite algorithms, pass the full key pair object from `generateSigningKey`
 * as the `skOrKeyPair` parameter.
 *
 * @param {number} algId
 * @param {Uint8Array|object} skOrKeyPair - Secret key bytes (pure) or full key pair object (composite)
 * @param {Uint8Array} handshakeHash - Transcript hash
 * @returns {Uint8Array} Signature bytes
 */
export function signHandshake(algId, skOrKeyPair, handshakeHash) {
  if (isComposite(algId)) {
    // For composite, skOrKeyPair should be the full key pair with _compositeKP
    const kp = skOrKeyPair._compositeKP || skOrKeyPair;
    return compositeSign(kp, handshakeHash);
  }
  const params = mldsaParams(algId);
  return mldsaSign(skOrKeyPair, handshakeHash, params);
}

/**
 * Verify a TLS 1.3 CertificateVerify signature.
 * @param {number} algId
 * @param {Uint8Array} pk - Public key
 * @param {Uint8Array} handshakeHash
 * @param {Uint8Array} signature
 * @returns {boolean}
 */
export function verifyHandshake(algId, pk, handshakeHash, signature) {
  if (isComposite(algId)) {
    const scheme = compositeScheme(algId);
    return compositeVerify(scheme, pk, handshakeHash, signature);
  }
  const params = mldsaParams(algId);
  return mldsaVerify(pk, handshakeHash, signature, params);
}

File: rust/pqc-tls/src/sig_algorithms.rs
//! PQC Signature Algorithms for TLS 1.3.
//!
//! Defines PQC and composite signature algorithms for the `signature_algorithms`
//! extension (CertificateVerify), along with sign/verify helpers.

extern crate alloc;
use alloc::vec::Vec;
use rand_core::{CryptoRng, RngCore};

use ml_dsa::dsa::{keygen as mldsa_keygen, sign as mldsa_sign, verify as mldsa_verify};
use ml_dsa::params::{MlDsa44, MlDsa65, MlDsa87};
use composite_sig::composite_sig::{
    CompositeScheme, CompositeKeyPair, CompositeSig,
    key_gen as composite_keygen, sign as composite_sign, verify as composite_verify,
    MLDSA65_ED25519, MLDSA87_ED25519,
};

/// TLS 1.3 signature algorithm identifiers for PQC.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u16)]
pub enum SignatureAlgorithm {
    /// ML-DSA-44 (code point 0x0904).
    MlDsa44 = 0x0904,
    /// ML-DSA-65 (code point 0x0905).
    MlDsa65 = 0x0905,
    /// ML-DSA-87 (code point 0x0906).
    MlDsa87 = 0x0906,
    /// ML-DSA-65 + Ed25519 composite (code point 0x0907).
    MlDsa65Ed25519 = 0x0907,
    /// ML-DSA-87 + Ed25519 composite (code point 0x0908).
    MlDsa87Ed25519 = 0x0908,
}

impl SignatureAlgorithm {
    /// Return the code point as a u16.
    pub fn code_point(self) -> u16 {
        self as u16
    }

    /// Look up a signature algorithm by its TLS code point.
    pub fn from_code_point(cp: u16) -> Option<Self> {
        match cp {
            0x0904 => Some(Self::MlDsa44),
            0x0905 => Some(Self::MlDsa65),
            0x0906 => Some(Self::MlDsa87),
            0x0907 => Some(Self::MlDsa65Ed25519),
            0x0908 => Some(Self::MlDsa87Ed25519),
            _ => None,
        }
    }

    /// All defined signature algorithms.
    pub const ALL: [SignatureAlgorithm; 5] = [
        Self::MlDsa44,
        Self::MlDsa65,
        Self::MlDsa87,
        Self::MlDsa65Ed25519,
        Self::MlDsa87Ed25519,
    ];

    /// Human-readable name.
    pub fn name(self) -> &'static str {
        match self {
            Self::MlDsa44 => "MLDSA44",
            Self::MlDsa65 => "MLDSA65",
            Self::MlDsa87 => "MLDSA87",
            Self::MlDsa65Ed25519 => "MLDSA65_ED25519",
            Self::MlDsa87Ed25519 => "MLDSA87_ED25519",
        }
    }

    /// Whether this is a composite (hybrid) signature algorithm.
    pub fn is_composite(self) -> bool {
        matches!(self, Self::MlDsa65Ed25519 | Self::MlDsa87Ed25519)
    }
}

fn composite_scheme(alg: SignatureAlgorithm) -> CompositeScheme {
    match alg {
        SignatureAlgorithm::MlDsa65Ed25519 => MLDSA65_ED25519,
        SignatureAlgorithm::MlDsa87Ed25519 => MLDSA87_ED25519,
        _ => panic!("not a composite algorithm"),
    }
}

/// A signing key pair for a PQC signature algorithm.
pub struct SigningKeyPair {
    /// Public key bytes.
    pub pk: Vec<u8>,
    /// Secret key bytes.
    pub sk: Vec<u8>,
    /// The algorithm this key pair is for.
    pub algorithm: SignatureAlgorithm,
}

/// Generate a signing key pair for the given signature algorithm.
pub fn generate_signing_key(
    alg: SignatureAlgorithm,
    rng: &mut (impl CryptoRng + RngCore),
) -> SigningKeyPair {
    match alg {
        SignatureAlgorithm::MlDsa44 => {
            let (pk, sk) = mldsa_keygen::<MlDsa44>(rng);
            SigningKeyPair { pk, sk, algorithm: alg }
        }
        SignatureAlgorithm::MlDsa65 => {
            let (pk, sk) = mldsa_keygen::<MlDsa65>(rng);
            SigningKeyPair { pk, sk, algorithm: alg }
        }
        SignatureAlgorithm::MlDsa87 => {
            let (pk, sk) = mldsa_keygen::<MlDsa87>(rng);
            SigningKeyPair { pk, sk, algorithm: alg }
        }
        SignatureAlgorithm::MlDsa65Ed25519 | SignatureAlgorithm::MlDsa87Ed25519 => {
            let scheme = composite_scheme(alg);
            let kp = composite_keygen(scheme, rng);
            SigningKeyPair { pk: kp.pk, sk: kp.sk, algorithm: alg }
        }
    }
}

/// Sign a TLS 1.3 CertificateVerify handshake hash.
///
/// The `handshake_hash` is the transcript hash that forms the content
/// of the CertificateVerify message.
pub fn sign_handshake(
    alg: SignatureAlgorithm,
    sk: &[u8],
    handshake_hash: &[u8],
) -> Vec<u8> {
    match alg {
        SignatureAlgorithm::MlDsa44 => mldsa_sign::<MlDsa44>(sk, handshake_hash),
        SignatureAlgorithm::MlDsa65 => mldsa_sign::<MlDsa65>(sk, handshake_hash),
        SignatureAlgorithm::MlDsa87 => mldsa_sign::<MlDsa87>(sk, handshake_hash),
        SignatureAlgorithm::MlDsa65Ed25519 | SignatureAlgorithm::MlDsa87Ed25519 => {
            let scheme = composite_scheme(alg);
            let kp = CompositeKeyPair {
                pk: Vec::new(),
                sk: sk.to_vec(),
                scheme,
            };
            let sig = composite_sign(&kp, handshake_hash);
            sig.bytes
        }
    }
}

/// Verify a TLS 1.3 CertificateVerify signature.
pub fn verify_handshake(
    alg: SignatureAlgorithm,
    pk: &[u8],
    handshake_hash: &[u8],
    signature: &[u8],
) -> bool {
    match alg {
        SignatureAlgorithm::MlDsa44 => mldsa_verify::<MlDsa44>(pk, handshake_hash, signature),
        SignatureAlgorithm::MlDsa65 => mldsa_verify::<MlDsa65>(pk, handshake_hash, signature),
        SignatureAlgorithm::MlDsa87 => mldsa_verify::<MlDsa87>(pk, handshake_hash, signature),
        SignatureAlgorithm::MlDsa65Ed25519 | SignatureAlgorithm::MlDsa87Ed25519 => {
            let scheme = composite_scheme(alg);
            let sig = CompositeSig { bytes: signature.to_vec() };
            composite_verify(scheme, pk, handshake_hash, &sig)
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::rngs::OsRng;

    #[test]
    fn test_signature_algorithm_code_points() {
        assert_eq!(SignatureAlgorithm::MlDsa44.code_point(), 0x0904);
        assert_eq!(SignatureAlgorithm::MlDsa65.code_point(), 0x0905);
        assert_eq!(SignatureAlgorithm::MlDsa87.code_point(), 0x0906);
        assert_eq!(SignatureAlgorithm::MlDsa65Ed25519.code_point(), 0x0907);
        assert_eq!(SignatureAlgorithm::MlDsa87Ed25519.code_point(), 0x0908);
    }

    #[test]
    fn test_signature_algorithm_from_code_point() {
        assert_eq!(
            SignatureAlgorithm::from_code_point(0x0905),
            Some(SignatureAlgorithm::MlDsa65)
        );
        assert_eq!(SignatureAlgorithm::from_code_point(0xFFFF), None);
    }

    #[test]
    fn test_mldsa65_sign_verify() {
        let kp = generate_signing_key(SignatureAlgorithm::MlDsa65, &mut OsRng);
        let hash = b"test handshake transcript hash for CertificateVerify";
        let sig = sign_handshake(SignatureAlgorithm::MlDsa65, &kp.sk, hash);
        assert!(verify_handshake(SignatureAlgorithm::MlDsa65, &kp.pk, hash, &sig));
    }

    #[test]
    fn test_composite_mldsa65_ed25519_sign_verify() {
        let kp = generate_signing_key(SignatureAlgorithm::MlDsa65Ed25519, &mut OsRng);
        let hash = b"composite signature handshake hash";
        let sig = sign_handshake(SignatureAlgorithm::MlDsa65Ed25519, &kp.sk, hash);
        assert!(verify_handshake(SignatureAlgorithm::MlDsa65Ed25519, &kp.pk, hash, &sig));
    }

    #[test]
    fn test_wrong_key_fails_verification() {
        let kp1 = generate_signing_key(SignatureAlgorithm::MlDsa65, &mut OsRng);
        let kp2 = generate_signing_key(SignatureAlgorithm::MlDsa65, &mut OsRng);
        let hash = b"test hash";
        let sig = sign_handshake(SignatureAlgorithm::MlDsa65, &kp1.sk, hash);
        assert!(!verify_handshake(SignatureAlgorithm::MlDsa65, &kp2.pk, hash, &sig));
    }
}
